publishing-sdk/src/main/java/com/ably/tracking/publisher/workerqueue/workers/ConnectionCreatedWorker.kt

```
@@ -0,0 +1,76 @@
+            // Leave Ably channel.
+            val presenceData = properties.presenceData.copy()
+            return SyncAsyncResult(
+                asyncWork = {
+                    val result = ably.disconnect(trackable.id, presenceData)
+                    if (result.isSuccess) {
+                        ConnectionCreatedWorkResult.RemovalRequested(trackable, handler, result.isSuccess)
+                    } else {
+                        ConnectionCreatedWorkResult.RemovalRequested(
+                            trackable,
+                            handler,
+                            result.isSuccess,
+                            result.exceptionOrNull() as ConnectionException?
+                        )
+                    }
+                }
+            )
+        }
+        val presenceData = properties.presenceData.copy()
+        return SyncAsyncResult(
+            asyncWork = {
+                subscribeToPresenceMessages(presenceData)
+            }
+        )
+    }
+
+    private suspend fun subscribeToPresenceMessages(presenceData: PresenceData): ConnectionCreatedWorkResult {
+        return suspendCoroutine { continuation ->
+            ably.subscribeForPresenceMessages(
+                trackableId = trackable.id,
+                listener = presenceUpdateListener,
+                callback = { result ->
+                    try {
+                        result.getOrThrow()
+                        continuation.resume(
+                            ConnectionCreatedWorkResult.PresenceSuccess(trackable, handler, presenceUpdateListener)
+                        )
+                    } catch (exception: ConnectionException) {
+                        ably.disconnect(trackable.id, presenceData) {
+                            continuation.resume(ConnectionCreatedWorkResult.PresenceFail(trackable, handler, exception))
+                        }
+                    }
+                }
```

I think that too much is happening here, instead of returning the `WorkResult` I'd suggest simply returning a success or failure and then using it to perform more actions. Thanks to this you'll be, for example, able to use suspending `disconnect` instead of the callback-based one :wink:

---

Good point 👍  However as in others I intentionally kept the async block as it's original implementation. I am going to add an issue to address this later

---

https://github.com/ably/ably-asset-tracking-android/issues/554